Implement the process for moving packages from testing to stable

matteodelabre opened the pull request:
This creates the `docs/branches.md` document which explains what the two release branches are, which one users should choose, and how the transition of packages from one branch to another is handled.

It also updates `docs/package.md` to add related information about the `maintainer` field.

This implements the proposal in #20, and closes #20 as a consequence.

matteodelabre committed:
Implement the process proposed in #20

matteodelabre commented:
This is mostly RFC, I tried to implement what we discussed, but please point out any problems in the documents.

Eeems commented:
@matteodelabre If this is a RFC right now, might as well convert it to a draft until we decide it's ready.

matteodelabre commented:
Not sure why the CI is failing since this only changes the docs.

Eeems commented:
> GitHub Actions has encountered an internal error when running your job.

Probably related to the Microsoft outage that's going on right now.

Eeems commented:
Just reviewed the content and it looks good to me. It would be nice to set up automation as part of this PR to create the merge window PR on a schedule as well.

raisjnn commented:
what's the process of becoming a maintainer and are the expectations of maintainers outlined somewhere?

matteodelabre commented:
> Just reviewed the content and it looks good to me. It would be nice to set up automation as part of this PR to create the merge window PR on a schedule as well.

Should this automatically-created PR contain all of the `package/` directory from the testing branch? I don’t think we can create empty PRs.

matteodelabre commented:
> what's the process of becoming a maintainer and are the expectations of maintainers outlined somewhere?

The process by which a maintainer is assigned to a package is described in this PR. As the document says, they are expected to review PRs regarding the packages they maintain. The process for adding a new repository maintainer is not outlined here, and I’d say it’s out of the scope of this PR.

Eeems commented:
> > Just reviewed the content and it looks good to me. It would be nice to set up automation as part of this PR to create the merge window PR on a schedule as well.
> 
> Should this automatically-created PR contain all of the `package/` directory from the testing branch? I don’t think we can create empty PRs.

I'd say no, likely we should just have a dummy file that we bump a timestamp in or something.

matteodelabre commented:
It turns out [we can actually create empty PRs](https://github.com/toltec-dev/test-empty-prs/pull/1), by adding an empty commit to a branch (with `git commit --allow-empty`).

matteodelabre committed:
Merge branch 'testing' into patch-specify-stable-testing

matteodelabre committed:
Merge branch 'testing' into patch-specify-stable-testing

matteodelabre commented on docs/branches.md:
```diff
@@ -0,0 +1,31 @@
+
+### Moving a package from testing to stable
+
+Each Monday, a pull request containing selected updates from the testing branch and targeting the stable branch can be opened.
+This pull request can only contain package changes dating from the previous Friday or older.
+Each of these package changes must be tested by a maintainer different from the maintainer of the affected package.
```
Instead of only allowing to open a merge PR on Mondays with changes from the preceding Friday, we could open them on Saturdays and only merge starting from the following Monday. This would allow us to more easily review and test the changes over the week-end.

```suggestion
Each Saturday, a pull request can be opened with selected updates from the testing branch to merge in the stable branch.
This pull request can only be merged from the following Monday.
Each of these package changes must be tested by a maintainer different from the maintainer of the affected package.
```

matteodelabre commented on docs/branches.md:
```diff
@@ -0,0 +1,31 @@
+
+Here are important things to check when testing a package:
+
+1. The package should work as intended by upstream.
+2. It must not destroy any user data (e.g. from previous versions of the same package, from other packages, from the home directory).
+3. It must not break other packages.
```
Some additional testing criteria that I suggest, following what we learned in the previous merge window.

```suggestion
1. It should work as intended by upstream.
2. It must not have any known major bugs.
3. It must not destroy any user data (e.g. from previous versions of the same package, from other packages, from the home directory).
4. It must not break other packages.
5. It must not lock the device in a state where it can not be used without troubleshooting from SSH.
```

raisjn replied:
> 5. It must not lock the device in a state where it can not be used without troubleshooting from SSH.

just as a side note, i thought of another workaround last week for the issue with the device going into bad state: the device can usually be rebooted, so if the device gets into a bad state while away from your computer, you can still get out of the bad state by rebooting.

matteodelabre replied:
That’s right. So maybe `5. It must not lock the device in a state where it can not be used without rebooting or troubleshooting from SSH.`?

raisjn replied:
> 2. It must not have any known major bugs.

I'd like it if we addressed how fixes can be brought over to stable during the merge window. Based on the last merge, if a package of mine gets punted to the next merge, I'm unhappy compared to if a fix can be brought in during the testing window. I'm not sure if others feel this way

matteodelabre committed:
Open merge PRs on Saturdays instead of Mondays

matteodelabre commented on docs/branches.md:
```diff
@@ -0,0 +1,31 @@
+
+### Moving a package from testing to stable
+
+Each Saturday, a pull request can be opened with selected updates from the testing branch to merge in the stable branch.
```
```suggestion
Each Saturday, a pull request can be opened with selected updates from the testing branch to merge in the stable branch.
Only bug fixes may be added to existing merge PRs over the weekend.
```

matteodelabre replied:
To address concerns from @raisjnn in the conversation above (https://github.com/toltec-dev/toltec/pull/81#discussion_r502918648).

matteodelabre committed:
Additional package testing criteria

matteodelabre committed:
Fix typos

matteodelabre committed:
Merge branch 'testing' into patch-specify-stable-testing

matteodelabre committed:
Add clause for allowing bug fixes

matteodelabre commented:
I think it’s best if we move forward with stabilizing the actual merge process then implement automation in a future PR.

raisjn reviewed (dismissed):
looks good to me. thanks for the update around bugfixes!

matteodelabre committed:
Update link to stable/testing docs in README

matteodelabre committed:
Add info on how to switch branches

matteodelabre committed:
Merge branch 'testing' into patch-specify-stable-testing

matteodelabre commented:
Just needed to fix a link and add instructions for switching to `testing` before merging.

matteodelabre commented:
If you still agree with this change set, could you re-approve the PR @raisjnn? Your review was dismissed since I changed the link in the README and added instructions to switch to `testing`. Thanks!

raisjn reviewed (approved):
thanks for adding the note about bug fixes!
